wazero/instance.go

```diff
@@ -0,0 +1,315 @@
+		vm:      vm,
+		imports: map[string]wazero.HostModuleBuilder{},
+		module:  module,
+		lock:    sync.Mutex{},
+	}
+	ins.stopCond = sync.NewCond(&ins.lock)
+
+	for _, option := range options {
+		option(ins)
+	}
+
+	if _, err := wasi_snapshot_preview1.Instantiate(ctx, r); err != nil {
+		_ = r.Close(ctx)
+		panic(err)
+	}
+
+	return ins
+}
+
+func (w *Instance) GetData() interface{} {
+	return w.data
+}
+
+func (w *Instance) SetData(data interface{}) {
+	w.data = data
+}
+
+func (w *Instance) Acquire() bool {
+	w.lock.Lock()
+	defer w.lock.Unlock()
+
+	if !w.checkStart() {
+		return false
+	}
+
+	w.refCount++
+
+	return true
+}
+
+func (w *Instance) Release() {
+	w.lock.Lock()
+	w.refCount--
+
+	if w.refCount <= 0 {
+		w.stopCond.Broadcast()
+	}
+	w.lock.Unlock()
+}
+
+func (w *Instance) Lock(data interface{}) {
+	w.lock.Lock()
+	w.data = data
+}
+
+func (w *Instance) Unlock() {
+	w.data = nil
+	w.lock.Unlock()
+}
+
+func (w *Instance) GetModule() common.WasmModule {
+	return w.module
+}
+
+func (w *Instance) Start() error {
+	ctx := context.Background()
+
+	for n, b := range w.imports {
+		if _, err := b.Instantiate(ctx, w.vm.engine); err != nil {
+			w.vm.engine.Close(ctx)
+			return err
+		}
+		delete(w.imports, n)
+	}
+
+	ins, err := w.vm.engine.InstantiateModule(ctx, w.module.module, wazero.NewModuleConfig())
+	if err != nil {
+		w.vm.engine.Close(ctx)
+		return err
+	}
+
+	w.instance = ins
+
+	atomic.StoreUint32(&w.started, 1)
+
+	return nil
+}
+
+func (w *Instance) Stop() {
+	go func() {
+		w.lock.Lock()
+		for w.refCount > 0 {
+			w.stopCond.Wait()
+		}
+		w.vm.engine.Close(context.Background())
+		_ = atomic.CompareAndSwapUint32(&w.started, 1, 0)
+		w.lock.Unlock()
+	}()
+}
+
+// return true is Instance is started, false if not started.
+func (w *Instance) checkStart() bool {
+	return atomic.LoadUint32(&w.started) == 1
+}
+
+func (w *Instance) RegisterFunc(namespace string, funcName string, f interface{}) error {
+	if w.checkStart() {
+		return ErrInstanceAlreadyStart
+	}
+
+	b, ok := w.imports[namespace]
+	if !ok {
+		b = w.vm.engine.NewHostModuleBuilder(namespace)
+		w.imports[namespace] = b
+	}
+
+	b.NewFunctionBuilder().WithFunc(f).Export(funcName)
+
+	return nil
+}
+
+func (w *Instance) Malloc(size int32) (uint64, error) {
+	if !w.checkStart() {
+		return 0, ErrInstanceNotStart
+	}
+
+	malloc, err := w.GetExportsFunc("malloc")
+	if err != nil {
+		return 0, err
+	}
+
+	addr, err := malloc.Call(size)
+	if err != nil {
+		w.HandleError(err)
+		return 0, err
+	}
+
+	return uint64(addr.(int32)), nil
+}
+
+func (w *Instance) GetExportsFunc(funcName string) (common.WasmFunction, error) {
+	if !w.checkStart() {
+		return nil, ErrInstanceNotStart
+	}
+
+	f := w.instance.ExportedFunction(funcName)
+	return &wasmFunction{f}, nil
+}
+
+type wasmFunction struct {
+	fn api.Function
+}
+
+// Call implements common.WasmFunction
+func (f *wasmFunction) Call(args ...interface{}) (interface{}, error) {
+	realArgs := make([]uint64, 0, len(args))
```

Comment:
this is the second direction of reflection. Not only are the function imports defined reflectively, they are also accessed the other way in reflectively. I think the intent was for this to be a generalized swappable runtime, but only one runtime lived here until now. This is the point I'd suggest being direct instead. make it a single runtime and then you don't need to do things reflectively. doing so would couple to wazero, but also make things a lot faster and also get back go context (as well drop all the system dependencies). worth a thought!